### ISSUE TITLE
File Manager

### ISSUE THREAD

**Issue by eurobank** [CIDRAM/CIDRAM#141]
Please change the default action of file manager from DELETE to something else (Edit maybe).

It is VERY easy to delete something right now.

**Comment by Maikuolan** [comment]
Good idea. Will work on this shortly. :+1:

**Comment by DanielRuf** [comment]
We have a filemanager? What exactly does it do and does it sanitize and check any input against RCE, path traversal and other things?

Generally files should be only managed through client software.

**Comment by DanielRuf** [comment]
Well, pretty big files (we should separate them into smaller ones).

https://github.com/CIDRAM/CIDRAM/blob/v2/vault/frontend_functions.php#531

Not sure if this is the best solution.
Also it makes sense to write some tests for this and add some comments to describe what the code does (some lines of this function are not self-explanatory).

If the filemanager disabled by default?

**Comment by Maikuolan** [comment]
> If the filemanager disabled by default?

It doesn't have it's own switch/configuration, but it's an integrated part of the front-end, so, whether enabled/disabled, depends on whether the front-end is enabled/disabled. As the front-end is disabled by default, the file manager is *effectively* disabled by default.

**Comment by Maikuolan** [comment]
> We have a filemanager? What exactly does it do

Some screenshots attached (screenshots are reasonably self-explanatory for what the file manager does; run-of-the-mill purpose, capabilities, etc, but specifically intended for working with the contents of the vault, so, can't be *directly* leveraged for working with anything above/outside the vault).

![Untitled](https://user-images.githubusercontent.com/12571108/61847679-22b87f00-aede-11e9-825c-3ad12db8deb1.png)
![Untitled2](https://user-images.githubusercontent.com/12571108/61847986-1e409600-aedf-11e9-9a7b-896fd8519210.png)
![Untitled3](https://user-images.githubusercontent.com/12571108/61847988-1e409600-aedf-11e9-8fee-c881e724fda9.png)

> and does it sanitize and check any input against RCE, path traversal and other things?

Yep. Paths are protected against traversal, RCE shouldn't be possible via the front-end itself, etc. However, if someone with admin access to the front-end *intentionally* tries to break their own installation, by deleting important files, by uploading compromised code or similar, that's on them (which is why I mentioned "directly"; could be *indirectly* possible if someone with access chooses to upload something to make it that way).

TL;DR – Quoting a few lines from [parts 4.0 and 4.1](https://github.com/CIDRAM/Docs/blob/master/readme.en.md#40-what-is-the-front-end) of the documentation, marking the super important bits (relevant to the questions here) in __*bold and italics*__:

> 4.0 WHAT IS THE FRONT-END.
>
> The front-end provides a convenient and easy way to maintain, manage, and update your CIDRAM installation. You can view, share, and download logfiles via the logs page, you can modify configuration via the configuration page, you can install and uninstall components via the updates page, and __*you can upload, download, and modify files in your vault via the file manager*__.
>
> The front-end is disabled by default in order to prevent unauthorised access (__*unauthorised access could have significant consequences for your website and its security*__). Instructions for enabling it are included below this paragraph.
>
> 4.1 HOW TO ENABLE THE FRONT-END.
>
> *(snip)*
>
> Note: After you've logged in for the first time, in order to prevent unauthorised access to the front-end, you should immediately change your username and password! This is very important, __*because it's possible to upload arbitrary PHP code to your website via the front-end*__.

(Assuming, of course, that the front-end has been enabled).

Also why front-end access should never be given to any parties that aren't 100% trusted (and, when it doubt, disabling the front-end entirely is always an option).

**Comment by Maikuolan** [comment]
This is also one of the reasons why three distinct permission levels are available for front-end accounts ("complete access", "logs access only", and "Cronable"). If someone wants to share the logs generated by their CIDRAM installation with a third party, directly via the front-end (e.g., situations like *"hey, check out my logs; here's access to my front-end so that you can see them directly for yourself, because I'm too lazy to send you the files myself"* or whichever similar situations, maybe for debugging purposes, to fix false positives, etc), they can create an account with "logs access only" (in which case, the only front-end pages accessible will be the logs page and the homepage).

![acc](https://user-images.githubusercontent.com/12571108/61849229-2dc1de00-aee3-11e9-9b76-b820c8cf5a24.png)

But, I would still generally advise against providing access to non-trusted parties anyhow, because, even with distinct permission levels available, providing any kind of access at all would still expose knowledge about the location of CIDRAM's main access endpoint (i.e., where the `loader.php` is located, and therefore where the vault is located). Knowing that information shouldn't generally pose any kind of threat, and obscurity is no kind of valid security in any modern environment anyway, but even so, better to not needlessly provide more information than necessary about the kinds of software/scripts/security/CMS/etc being used to non-trusted parties.

But, it's there, in case the distinct permission levels are needed by anyone. (The "Cronable" permission level acts quite differently than the other two permission levels, and should only ever be used for accounts that Cronable uses; it can access/leverage the updates page, and a few other things that Cronable might need, but nothing else; also, if a human somehow logs in using an account with that permission level, everything will look broken, because template data is rendered differently when that permission level is in use).

**Comment by ghost** [comment]
> > We have a filemanager? What exactly does it do

Welcome to the world :grin: 

> TL;DR – Quoting a few lines from [parts 4.0 and 4.1](https://github.com/CIDRAM/Docs/blob/master/readme.en.md#40-what-is-the-front-end) of the documentation, marking the super important bits (relevant to the questions here) in **_bold and italics_**:

No matter how much time we put into writing copious amounts of warnings, there will always be someone who persists on not reading them for any imaginable reason and then screws something up completely. Universe has an impressive way to creating bigger and more brain dead fools the more time passes and knowledge levels rises. And despite being served with proper education people still lacks common sense when it comes to the most basic functions of anything... 

TL:DR we end up with too long manuals that no one will read because of the length.

**Comment by Maikuolan** [comment]
Very true. Only so much we can do about the brain-dead types of people out there though, unfortunately.

**Comment by DanielRuf** [comment]
We should split the docs then (CLI and frontend).

**Comment by Maikuolan** [comment]
Implemented via b408d73c6e82b16c03a5806cc44fa1b96b1dfae6.

**Comment by Maikuolan** [comment]
> Well, pretty big files (we should separate them into smaller ones).
>
> https://github.com/CIDRAM/CIDRAM/blob/v2/vault/frontend_functions.php#531
>
> Not sure if this is the best solution.
> Also it makes sense to write some tests for this and add some comments to describe what the code does (some lines of this function are not self-explanatory).

> We should split the docs then (CLI and frontend).

Definitely, there is room for improvement. Probably better to break things up a little before writing more tests though, so that writing the tests will be easier (can call the specific needed files and such). But anyway, let's create separate issues for these (so that I can close this issue off, now that the original suggestion has been implemented). :-)
